Remove the wrong codegen assumption that a conversion between non-byte calldata arrays can never happen (#16415)
ethereum/solidity: Changelog.md

* Yul EVM Code Transform: Improve stack shuffler performance by fixing a BFS deduplication issue.

Bugfixes:
* TypeChecker: Remove the wrong codegen assumption that a conversion between non-byte calldata arrays can never happen.

> * TypeChecker: Fix tuple components types comparison for non-trivial types. This bug led to unnecessary conversions between tuples of the same types.

Reviewer: ```suggestion
* TypeChecker: Fix spurious comparison of tuples containing reference types, which led to unnecessary conversions between tuples of the same type.
```
Just a suggestion.
But I think that "non-trivial types" should actually be "reference types", right?

Author: I think you are right but maybe someone should take an additional look. Not sure if there are only reference types.

* Yul: Fix incorrect serialization of Yul object names containing double quotes and escape sequences, producing output that could not be parsed as valid Yul.
* Yul EVM Code Transform: Improve stack shuffler performance by fixing a BFS deduplication issue.
* Yul IR Code Generation: Preserve custom error argument of `require` when stripping of revert strings is selected via `--revert-strings strip`.
* TypeChecker: Remove the wrong codegen assumption that a conversion between non-byte calldata arrays can never happen.

> * TypeChecker: Fix spurious comparison of tuples containing reference types, which led to unnecessary conversions between tuples of the same type.

Reviewer 2: This needs to be moved to 0.8.36.

Also, not sorted alphabetically.

Reviewer 2: Category is wrong - both are IR codegen bugs, not analysis bugs. They could also use a bit more detail from syntax perspective. I did not understand the second one myself until I saw the code.

There's also the question of whether it only affected reference types. I guess it depends on which types are guaranteed to have one instance in type provider. If you're unsure, you can also just not specify that.

```suggestion
* Yul IR Code Generation: Fix internal compiler error when non-byte calldata arrays of the same type were involved in operations that require converting operands to a common type, such as ternary operator.
* Yul IR Code Generation: Fix identical tuple types sometimes being seen by the compiler as different, leading to superfluous conversion code being emitted.
```

To make sure I actually do understand it now: the problem with the second one was that we sometimes had different `Type` objects for the same reference type and they compared as different even though they should have been seen as the same? And this resulted in `arrayConversionFunction()` being used unnecessarily?
